### PR TITLE
Public sharing

### PR DESCRIPTION
**Pull request by georgehrke**

### CODE REVIEW

**Comment by georgehrke** [comment]
@tcitworld

**Comment by raghunayyar** [comment]
Can you add the steps to check?
This is one `massive` Pull Request. :)

**Comment by tcitworld** [comment]
@raghunayyar 
- [ ] Check that it doesn't break anything without the server change : https://github.com/nextcloud/server/pull/1197 (i.e work on NC < 11).
- [ ] Test that the features comes correctly with the server change
- [ ] Test publish link is accessible and contains all events
- [ ] Test email is correctly sent with link

Todos : 
- [ ] More tests can be added
- [ ] OpenGraph informations on a link

**Comment by georgehrke** [comment]
pr in server: https://github.com/nextcloud/server/pull/1197

**Comment by georgehrke** [comment]
@tcitworld Can you take care of rebasing? :)

**Comment by tcitworld** [comment]
How did you know ? I was doing it right now !

**Comment by georgehrke** [comment]
hehe 🙈

**Comment by georgehrke** [comment]
https://github.com/nextcloud/server/pull/1197 was merged, so let's get this in by end of the week :)

**Comment by tcitworld** [comment]
It's the end of the week. ;-) Not sure if 856d761 works properly.

**Comment by georgehrke** [comment]
nextcloud chromium today at 2 53 26 pm
I'm not too happy, that these are in the settings.
It's not really settings and has a bad discoverability I think.
But we can take care of this in a follow-up PR. not a blocker

**Comment by georgehrke** [comment]
- [ ] please only add the public graph thing to main.php when it's a public page

**Comment by georgehrke** [comment]
calendar - nextcloud chromium today at 3 24 47 pm
Send button jumps into next line when list is scrollable

File: controller/viewcontroller.php [hunk 1]
```diff
@@ -155,4 +231,57 @@ private function getTimezoneList() {
...
+			return new JSONResponse([], Http::STATUS_BAD_REQUEST);
+		}
+
+		$sendFromDomain = $this->config->getSystemValue('mail_domain', 'domain.org');
+		$sendFromAddress = $this->config->getSystemValue('mail_from_address', 'owncloud');
+		$sendFrom = $sendFromAddress . '@' . $sendFromDomain;
+
+		$message = $this->mailer->createMessage();
+		$message->setSubject($subject);
+		$message->setFrom([$sendFrom => 'ownCloud Notifier']);
```
Comment (georgehrke) [inline controller/viewcontroller.php]: 🙈

File: controller/viewcontroller.php [hunk 2]
```diff
@@ -155,4 +231,57 @@ private function getTimezoneList() {
...
+		$message->setTo([$target => 'Recipient']);
+		if ($useHTML) {
+			$message->setHtmlBody($body);
+		} else {
+			$message->setPlainBody($body);
+		}
+		$this->mailer->send($message);
+
+		return new JSONResponse([], Http::STATUS_OK);
```
Comment (georgehrke) [inline controller/viewcontroller.php]: the `JSONResponse` call should be moved to `sendEmailPublicLink`

**Comment by georgehrke** [comment]
Right now we are only sending an html-body. Please also provide a plain text message.
see http://swiftmailer.org/docs/messages.html#setting-the-body-content

We should not force users to use html emails.

**Comment by georgehrke** [comment]
Looking at the source code of OCA\Mail\Message, I suppose that you first have to use `setPlainBody` and afterwords set the html content with `setHtmlBody`

https://github.com/nextcloud/server/blob/8c7d7d7746e76b77ad86cee3aae5dbd4d1bcd896/lib/private/Mail/Message.php#L211

File: templates/part.calendarlist.item.php [hunk 3]
```diff
@@ -205,4 +206,28 @@ class="app-navigation-entry-menu hidden">
 			</span>
 		</li>
 	</ul>
+	<div class="publishing" ng-if="item.calendar.isPublishable()">
+		<input type="checkbox" name="publish"
+			   id="checkbox_publish_calendar_{{ $index }}"
+			   ng-model="item.calendar.published" value="edit"
+			   ng-change="togglePublish(item)">
+		<label for="checkbox_publish_calendar_{{ $index }}"> <?php p($l->t('Share link')); ?></label>
```
Comment (georgehrke) [inline templates/part.calendarlist.item.php]: please use the fancy checkbox styles from the server :) 
see https://github.com/nextcloud/calendar/pull/79

**Comment by georgehrke** [comment]
thx! will test again asap

**Comment by tcitworld** [comment]
@georgehrke Fixed.

> I'm not too happy, that these are in the settings.
It's not really settings and has a bad discoverability I think.

Agreed. Let's discuss this with @jancborchardt somewhere.

**Comment by tcitworld** [comment]
- [ ] Fix inferface urls ending with `#` causing bad redirection

**Comment by jancborchardt** [comment]
@georgehrke @tcitworld mind posting some quick screenshots for easier review? :) My calendar dev environment is also a bit broken atm, will open issue.

**Comment by georgehrke** [comment]
> My calendar dev environment is also a bit broken atm, will open issue.

Please try `make distclean && make` first. :)
It will delete every production and development dependencies and refetch them.

**Comment by georgehrke** [comment]
The mail contains html as plain text:
```
--_=_swift_v4_1476366921_7cee9c87b934b295a3d7eff5c1f45643_=_
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<!DOCTYPE html>
<html class=3D"ng-csp" data-placeholder-focus=3D"false" l=
ang=3D"en" >
=09<head data-requesttoken=3D"LREmSBUsJzshCRsKPCkzU1kvXWAiLm=
NAaQswClkEBgA=3D:iTpqvYCsuzaeUsVxjx4OQzLsZluLvO1eEOWlo5Hmpjw=3D">
=09=09<=
meta charset=3D"utf-8">
=09=09<title>
=09=09Nextcloud=09=09</title>
=
=09=09<meta http-equiv=3D"X-UA-Compatible" content=3D"IE=3Dedge">
=09=09<=
meta name=3D"referrer" content=3D"never">
=09=09<meta name=3D"viewport" c=
ontent=3D"width=3Ddevice-width, minimum-scale=3D1.0, maximum-scale=3D1.0">=
```
Will look into it.

File: templates/mail.publication.text.php [hunk 4]
```diff
@@ -1,13 +1,11 @@
 <?php p($l->t('Hello,')); ?>
-<?php p('\n\n'); ?>
 
-<?php $l->t("We wanted to inform you that %s has published the calendar %s.", [$_['username'],$_['calendarname']]); ?>
-<?php p('\n'); ?>
 
-<?php p($l->t('Click on the link below to access it !')); ?>
-<?php p('\n'); ?>
```
Comment (georgehrke) [inline templates/mail.publication.text.php]: @tcitworld Just as a side note:
When using `\n` in PHP you have to you doubles quotes. `'\n'` won't work.
And as this is a template, you can just insert actual new lines :)

**Comment by georgehrke** [comment]
@tcitworld Have you tested the public graph thing in production yet?
framagenda seems to use it already 😉

File: templates/mail.publication.text.php (hunk 4 above)
Comment (tcitworld) [inline templates/mail.publication.text.php]: Yeah, I noticed this a few days ago. :-)

**Comment by tcitworld** [comment]
Yup. For example, [this one](https://framagenda.org/index.php/apps/calendar/public/TS9MXY3HXCKBLOFA) works on http://iframely.com/debug and on Facebook, though Facebook doesn't seem to fetch the thumbnail.

Overkill to add twitter cards ?

**Comment by georgehrke** [comment]
> Yup. For example, this one works on http://iframely.com/debug and on Facebook, though Facebook doesn't seem to fetch the thumbnail.

I think we can investigate the fb issue later.
Maybe the thumbnail we send is too small or something.
Definitely not a blocker.

> Overkill to add twitter cards ?

Maybe later. This pull request is already huge ^^

**Comment by georgehrke** [comment]
Okay, as the public graph thing works, this is a 👍 for me :)

**Comment by raghunayyar** [comment]
I think its high time we get this in. :+1:

**Comment by georgehrke** [comment]
thx a lot @tcitworld! :)

**Comment by jancborchardt** [comment]
@tcitworld very very nice! :) :tada:

**Pull request merged**